service/src/builder.rs

```diff
@@ -209,64 +240,52 @@ async fn compile(
     project_path: PathBuf,
     target_path: impl Into<PathBuf>,
     deployment: bool,
-    tx: Sender<Message>,
+    tx: tokio::sync::mpsc::Sender<String>,
 ) -> anyhow::Result<Vec<BuiltService>> {
     let manifest_path = project_path.join("Cargo.toml");
     if !manifest_path.exists() {
-        bail!("failed to read the Shuttle project manifest");
+        bail!("Cargo manifest file not found: {}", manifest_path.display());
     }
     let target_path = target_path.into();
 
-    let mut cargo = tokio::process::Command::new("cargo");
-    cargo
-        .arg("build")
+    let mut cmd = tokio::process::Command::new("cargo");
+    cmd.arg("build")
         .arg("--manifest-path")
         .arg(manifest_path)
         .arg("--color=always") // piping disables auto color, but we want it
         .current_dir(project_path.as_path());
 
     if deployment {
-        cargo.arg("--jobs=4");
+        cmd.arg("--jobs=4");
     }
 
     for package in &packages {
-        cargo.arg("--package").arg(package.name.as_str());
+        cmd.arg("--package").arg(package.name.as_str());
     }
 
     let profile = if release_mode {
-        cargo.arg("--release");
+        cmd.arg("--release");
         "release"
     } else {
         "debug"
     };
 
     if wasm {
-        cargo.arg("--target").arg("wasm32-wasi");
+        cmd.arg("--target").arg("wasm32-wasi");
     }
 
-    let (reader, writer) = os_pipe::pipe()?;
-    let writer_clone = writer.try_clone()?;
-    cargo.stdout(writer);
-    cargo.stderr(writer_clone);
-
-    let mut handle = cargo.spawn()?;
-
-    tokio::task::spawn_blocking(move || {
-        let reader = std::io::BufReader::new(reader);
-        for line in reader.lines() {
-            if let Ok(line) = line {
-                if let Err(error) = tx.send(Message::TextLine(line)) {
-                    error!("failed to send cargo message on channel: {error}");
-                };
-            } else {
-                error!("Failed to read Cargo log messages");
-            };
+    cmd.stderr(Stdio::piped());
+    cmd.stdout(Stdio::null());
+    let mut handle = cmd.spawn()?;
+    let reader = tokio::io::BufReader::new(handle.stderr.take().unwrap());
+    tokio::spawn(async move {
+        let mut lines = reader.lines();
+        while let Some(line) = lines.next_line().await.unwrap() {
+            let _ = tx.send(line).await.map_err(|e| error!("{e}"));
```

Comment:
Would you mind explaining this change? It seems to me like we are first ignoring the stderr stream, and then we start reading it (`handle.stderr.take()`)? Or is this a better way to combine stderr and stdout into one stream?

Reply:
I've changed this implementation to only read stderr, as I discovered that cargo build seems to only print to stderr. `deployer/src/deployment/queue.rs` is an example where I open two readers for reading both streams.

Reply:
Alright, thanks for the explanation!